pkg/converter/convert_unix.go

```diff
@@ -592,89 +592,130 @@ func LayerConvertFunc(opt PackOption) converter.ConvertFunc {
 
 // ConvertHookFunc returns a function which will be used as a callback
 // called for each blob after conversion is done. The function only hooks
-// the manifest conversion and merges all the nydus blob layers into a
-// nydus bootstrap layer and update the image config.
+// the index conversion and the manifest conversion.
 func ConvertHookFunc(opt MergeOption) converter.ConvertHookFunc {
 	return func(ctx context.Context, cs content.Store, orgDesc ocispec.Descriptor, newDesc *ocispec.Descriptor) (*ocispec.Descriptor, error) {
-		if !images.IsManifestType(newDesc.MediaType) {
-			// Only need to hook manifest conversion
+		switch {
+		case images.IsIndexType(newDesc.MediaType):
+			return convertIndex(ctx, cs, orgDesc, newDesc)
+		case images.IsManifestType(newDesc.MediaType):
+			return convertManifest(ctx, cs, newDesc, opt)
+		default:
 			return newDesc, nil
 		}
+	}
+}
 
-		// Convert manifest
-		var manifest ocispec.Manifest
-		manifestDesc := *newDesc
-		labels, err := readJSON(ctx, cs, &manifest, manifestDesc)
-		if err != nil {
-			return nil, errors.Wrap(err, "read manifest json")
-		}
-
-		// Append bootstrap layer to manifest.
-		bootstrapDesc, blobDescs, err := MergeLayers(ctx, cs, manifest.Layers, MergeOption{
-			BuilderPath:   opt.BuilderPath,
-			WorkDir:       opt.WorkDir,
-			ChunkDictPath: opt.ChunkDictPath,
-			WithTar:       true,
-		})
-		if err != nil {
-			return nil, errors.Wrap(err, "merge nydus layers")
+// convertIndex modifies the original index by appending "nydus.remoteimage.v1"
+// to the Platform.OSFeatures of each modified manifest descriptors.
+func convertIndex(ctx context.Context, cs content.Store, orgDesc ocispec.Descriptor, newDesc *ocispec.Descriptor) (*ocispec.Descriptor, error) {
+	var orgIndex ocispec.Index
+	if _, err := readJSON(ctx, cs, &orgIndex, orgDesc); err != nil {
+		return nil, errors.Wrap(err, "read target image index json")
+	}
+	// isManifestModified is a function to check whether the manifest is modified.
+	isManifestModified := func(manifest ocispec.Descriptor) bool {
```

Comment:
Maybe we can check whether it is nydus manifest (name it `isNydusManifest`) by recognizing bootstrap annotation, which would be safer?

Reply:
Then we need to read json from the content store for each manifest descriptor.

Reply:
Maybe we can store the modified manifest digest in a map in the manifest conversion hook?

Reply:
Well, just checking if the manifest is changed should also work.